roachtest: clean up allocate cluster function

RaduBerinde opened the pull request:
#### roachtest: remove clusterAllocator argument

The only reason for this argument was so we could inject errors in one
test. This commit replaces it with a "pre allocate" function in the
cluster opts.

Epic: none
Release note: None

#### roachtest: clean up allocate cluster function

Replace the allocate cluster function with a method. This makes things
more clear, especially the lifetime of the cluster factory.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111534)
<!-- Reviewable:end -->

srosenberg reviewed (approved):
Nice, thanks for refactoring `clusterAlloc`!

Since it touches nearly every (remote) test execution, I've triggered [`SELECT_PROBABILITY=0.4`](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/12040481?buildTab=log&focusLine=33&logView=linear).

RaduBerinde committed:
roachtest: remove clusterAllocator argument

The only reason for this argument was so we could inject errors in one
test. This commit replaces it with a "pre allocate" function in the
cluster opts.

Epic: none
Release note: None

RaduBerinde committed:
roachtest: clean up allocate cluster function

Replace the allocate cluster function with a method. This makes things
more clear, especially the lifetime of the cluster factory.

Epic: none
Release note: None

RaduBerinde commented:
bors r+
